### PR TITLE
Ignore the changes of status.conditions and useless annotations in svc during service registration

### LINKED ISSUE #44923: Ignore the changes of "status.conditions" in `svc` during service registration.
(This is used to request new product features, please visit <https://discuss.istio.io> for questions on using Istio)

**Describe the feature request**
Currently, any changes in the `conditions` field in `k8s svc` triggers istiod to push XDS to the data plane, which is unnecessary. Frequent changes in `conditions` will result in increased resource consumption by istiod, which is something that users do not want to see.

**Describe alternatives you've considered**
We should ignore the changes of "status.conditions" in `svc` during service registration.

**Affected product area (please put an X in all that apply)**

[ ] Ambient
[ ] Docs
[ ] Installation
[X ] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Affected features (please put an X in all that apply)**

[ ] Multi Cluster
[ ] Virtual Machine
[ ] Multi Control Plane

**Additional context**

### PR DESCRIPTION
**Pull request by miaoyq**
**Please provide a description of this PR:**

fix #44923

### CODE REVIEW

File: pilot/pkg/serviceregistry/kube/controller/controller.go [hunk 1]
```diff
@@ -408,6 +411,58 @@ func (c *Controller) onServiceEvent(_, curr *v1.Service, event model.Event) erro
 	return nil
 }
 
+func serviceEqual(first, second *v1.Service) bool {
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/controller.go:414]: I am very concerned about the cost

**Comment by hzxuzhonghu** [comment]
The original issue is saying conditions update.

File: pilot/pkg/serviceregistry/kube/controller/controller.go (hunk 1 above)
Comment (ramaraochavali) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:414]: +1

**Review by howardjohn — CHANGES_REQUESTED** [review]
Don't we already have logic to filter out unneeded service events? #42617?

**Comment by miaoyq** [comment]
> The original issue is saying conditions update.

@hzxuzhonghu But the change of annotations will make istiod handle the svc also, that is unnecessary.

**Comment by miaoyq** [comment]
> Don't we already have logic to filter out unneeded service events? #42617?

@howardjohn This PR only handled labels, did not handle annotations and conditions, and I think we can avoid unnecessary processing of istiod from the source.

File: pilot/pkg/serviceregistry/kube/controller/controller.go (hunk 1 above)
Comment (miaoyq) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:414]: Is the meaning that we shouldn't use `reflect` or something else?

Comment (miaoyq) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:418]: @hzxuzhonghu @ramaraochavali I have add `serviceEqual` to the beginning of `onServiceEvent`.  Please take a look again. Thx

Comment (ramaraochavali) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:418]: I think using reflect Equals is bad. If you want to handle conditions, can it be done here https://github.com/istio/istio/blob/161e45ef52fe6ab65f1bec5aad42476163a776f4/pilot/pkg/model/service.go#L1115

Comment (miaoyq) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:418]: @ramaraochavali  `Equals` is compare `model.Service`, that have't `conditions`.  In addition, it seems to only take effect after the code below?

Comment (ramaraochavali) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:418]: Annotations and labels are already part of the service comparison I shared above. If you want only conditions - I would say just compare conditions here similar to how the labels and annotations are compared there. But not sure how frequent the condition updates would be

Comment (miaoyq) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:418]: > If you want only conditions - I would say just compare conditions here similar to how the labels and annotations are compared there.

I'm not sure if I understand correctly, but as far as I know, there is no Conditions information in model.Service. Therefore, we must compare the Conditions field in the Kubernetes service. 

> But not sure how frequent the condition updates would be

The changes of `Node` configuration will affect the status of the `SVC` when the `SVC` is LB type , so some ServiceController implementations maybe write some information to the conditions of the `SVC`.

@howardjohn

**Comment by howardjohn** [comment]
> > Don't we already have logic to filter out unneeded service events? #42617?
> 
> @howardjohn This PR only handled labels and annotations, did not handle conditions, and I think we can avoid unnecessary processing of istiod from the source.

I don't see how it ignores conditions. model.Service doesn't have conditions, so any conditions change will not impact the equality check done (`prev.Equals(curr)`)

**Comment by miaoyq** [comment]
> So how can condition changing make this "miss"?

@howardjohn I apologize for any confusion I may have caused. What I meant to say is that, before handling the k8s svc, we can filter out any changes that may not be necessary in order to avoid any unnecessary processing.

For example, when there are changes in `conditions`, we do not need to convert `k8s.svc` to `model.svc` and not need to call `addOrUpdateService`.

**Comment by ramaraochavali** [comment]
> Do you care about conditions? If not, can we strip it calling .SetTransform

If this works, I Think this is better approach.

**Comment by miaoyq** [comment]
> Do you care about conditions? If not, can we strip it calling .SetTransform

@hzxuzhonghu  This doesn't seem to works. `SetTransform` can set `conditions` to `nil`, but can not prevent updating events.

**Comment by hzxuzhonghu** [comment]
That's unfortunate, i think the only way is to filter out unrelevant events in `onServiceEvent`. But keep in mind to be more efficient

**Comment by howardjohn** [comment]
Why do we care about the events? They do NOT result in a XDS push, right? Because we filter them later?

Even the original issue says "...triggers istiod to push XDS to the data plane, which is unnecessary..."

**Comment by hzxuzhonghu** [comment]
I think @miaoyq meant the handler will do some work like service convertion, which is a waste

**Comment by miaoyq** [comment]
> That's unfortunate, i think the only way is to filter out unrelevant events in onServiceEvent. But keep in mind to be more efficient

@hzxuzhonghu I have refactor the "Equal" code, PTAL again, Thx.

**Comment by ramaraochavali** [comment]
I also thought we are trying to solve XDS pushes as documented [here](https://github.com/istio/istio/issues/44923#issue-1711125727). If it is just events - I think it is not a huge problem. I do not think it is worth complicating this.

**Comment by miaoyq** [comment]
> I don't think we should do this. It is complicate - this is 800 lines of fragile code with likely no performance benefit.

@howardjohn Most of the code is test cases (620+ lines), but if this would increase the understanding and maintenance cost, I also agree with you. :-)

**Pull request closed without merging**